diff --git a/src/components/CellRendererComponent.tsx b/src/components/CellRendererComponent.tsx
index d7b833b7d70e43bfcec14b679cb9d43597b4a7d2..b9c2d0659d7bb3c4ed9197a7ed599c3e8077b71d 100644
--- a/src/components/CellRendererComponent.tsx
+++ b/src/components/CellRendererComponent.tsx
@@ -128,7 +128,7 @@ function CellRendererComponent<T>(props: Props<T>) {
     itemLayoutAnimation,
   } = propsRef.current;
 
-  useEffect(() => {
+  /* useEffect(() => {
     // NOTE: Keep an eye on reanimated LayoutAnimation refactor:
     // https://github.com/software-mansion/react-native-reanimated/pull/3332/files
     // We might have to change the way we register/unregister LayouAnimations:
@@ -149,7 +149,7 @@ function CellRendererComponent<T>(props: Props<T>) {
         global.LayoutAnimationRepository.registerConfig(t, stashedConfig);
       }
     })(tag, layoutAnimationDisabled);
-  }, [layoutAnimationDisabled]);
+  }, [layoutAnimationDisabled]); */
 
   return (
     <Animated.View
